File: packages/manifest-schema/src/manifest.ts
/**
 * Activation manifest schema -- spec v5 §11.2, §75.
 *
 * The manifest is the ONLY thing that authorises a Partner Agent to serve an
 * ad or to start an external upload. §11.2 states the rule plainly:
 *
 *   "No valid signature + no valid approval + no current policy = no ad and
 *    no external upload."
 *
 * Every field here exists because the Agent must be able to enforce the
 * Partner's decision offline, without calling back to Oolix (§75 stale grace).
 */
import { z } from 'zod';
import { ChannelSchema, FrequencyCapSchema, IsoDateTimeSchema } from '@oolix/contracts';

/** §75 JWS protected header type. */
export const MANIFEST_JWS_TYP = 'OOLIX-MANIFEST+JWS';

/** §75: manifests are signed ES256. No other algorithm is accepted. */
export const MANIFEST_ALG = 'ES256';

export const ManifestBudgetSchema = z.object({
  /**
   * Minor units as a string. §73 stores budgets as BIGINT; JSON numbers lose
   * precision past 2^53, and a truncated budget is a financial defect, so the
   * wire format is a decimal string.
   */
  allocation_minor: z.string().regex(/^\d+$/, 'decimal minor units'),
  currency: z.string().length(3),
  /**
   * §76.1: the Agent stops locally at this fraction of the allocation. The
   * reserve absorbs reporting lag so the Partner cannot overdeliver against a
   * budget the control plane has not yet confirmed.
   */
  local_stop_fraction: z.number().min(0).max(1).default(0.98),
});

export const ManifestPayloadSchema = z.object({
  manifest_version: z.number().int().nonnegative(),
  activation_id: z.string().min(1),
  partner_org_id: z.string().min(1),
  /**
   * v6 §19: legacy prebuilt-segment targeting.
   *
   * Nullable because the primary path is now an Audience Group. A Partner with
   * an existing CDP audience can still be activated this way, and an Agent in
   * the field must accept both shapes -- §80 keeps N-1 compatibility during a
   * rollout, so a manifest carrying either must verify.
   */
  segment_id: z.string().min(1).nullable().optional(),
  /** Partner-internal segment key the Agent resolves locally (§38.1). */
  segment_key: z.string().min(1).nullable().optional(),

  /**
   * v6 Appendix B: the approved audience.
   *
   * `rule_hash` is the load-bearing field. §11 has the Agent compile the rules
   * locally and materialize the matching members; recomputing the hash before
   * it does so is how it confirms it is about to serve exactly what the Partner
   * approved (§10) rather than something edited since.
   */
  audience: z
    .object({
      audience_group_id: z.string().min(1),
      audience_version: z.number().int().positive(),
      rule_hash: z.string().length(64),
      /** The rules themselves, so the Agent can compile without calling back. */
      rules: z
        .array(
          z.object({
            attribute: z.string().min(1),
            operator: z.enum(['EQ', 'IN', 'LTE', 'GTE', 'BETWEEN']),
            value: z.unknown(),
            required: z.boolean(),
            weight: z.number().int().min(1).max(5),
          }),
        )
        .min(1),
      reach_estimate_id: z.string().min(1).nullable().optional(),
      capability_version: z.number().int().positive().nullable().optional(),
      mapping_version: z.number().int().positive().nullable().optional(),
    })
    .nullable()
    .optional(),
  channel: ChannelSchema,
  placement_ids: z.array(z.string().min(1)),
  /** Partner-published placement keys, so the Agent matches without a lookup. */
  placement_keys: z.array(z.string().min(1)),
  creative_version_ids: z.array(z.string().min(1)).min(1),
  budget: ManifestBudgetSchema,
  frequency_cap: FrequencyCapSchema,
  /** §81: explicit, versioned purpose -- never a permanent eligible boolean. */
  purpose_id: z.string().min(1),
  policy_version: z.string().min(1),
  /** Category constraints the Agent enforces locally (§41). */
  allowed_categories: z.array(z.string()).default([]),
  blocked_categories: z.array(z.string()).default([]),
  campaign_category: z.string().min(1),
  issued_at: IsoDateTimeSchema,
  /**
   * §75: after this instant the Agent must stop serving from this manifest,
   * even if Oolix is unreachable. This is what makes the offline cache safe.
   */
  config_expires_at: IsoDateTimeSchema,
  start_at: IsoDateTimeSchema,
  end_at: IsoDateTimeSchema,
  /** Points at the approval row that authorised this (§83 auditability). */
  approval_reference: z.string().min(1),
  /** §40.4 / §41: audience expansion is off unless the Partner allowed it. */
  audience_expansion_allowed: z.boolean().default(false),
});

export type ManifestPayload = z.infer<typeof ManifestPayloadSchema>;
export type ManifestBudget = z.infer<typeof ManifestBudgetSchema>;

/**
 * Creative bundle delivered alongside the manifest.
 *
 * Kept OUT of the signed payload deliberately: creative assets are large and
 * cacheable, while the manifest must stay small enough to re-sign cheaply on
 * every budget or policy change. Integrity is preserved by binding each
 * creative to its content hash, which the Agent verifies against the asset it
 * actually fetched (§70: "Partner approval binds to the content hash").
 */
export const ManifestCreativeSchema = z.object({
  creative_version_id: z.string().min(1),
  type: z.enum(['IMAGE', 'NATIVE_CARD']),
  asset_url: z.string().url().nullable(),
  content_sha256: z
    .string()
    .regex(/^[0-9a-f]{64}$/, 'lowercase hex sha-256')
    .nullable(),
  width: z.number().int().positive().nullable(),
  height: z.number().int().positive().nullable(),
  headline: z.string().nullable(),
  body: z.string().nullable(),
  cta: z.string().nullable(),
  destination_url: z.string().url(),
  legal_disclaimer: z.string().nullable(),
});

export type ManifestCreative = z.infer<typeof ManifestCreativeSchema>;

/** What GET /agent/v1/config/pull returns (§52.4). */
export const AgentConfigBundleSchema = z.object({
  config_version: z.number().int().nonnegative(),
  issued_at: IsoDateTimeSchema,
  partner_org_id: z.string().min(1),
  /** Compact JWS strings, one per active activation. */
  manifests: z.array(z.string().min(1)),
  creatives: z.array(ManifestCreativeSchema),
  /** Activation IDs the Agent must stop serving immediately (§24). */
  revoked_activation_ids: z.array(z.string()),
  /** §24 / §56: partner-scoped kill switches the Agent enforces locally. */
  kill_switches: z.array(
    z.object({
      scope: z.enum(['AGENT', 'PLACEMENT', 'ACTIVATION', 'CHANNEL', 'PARTNER_ALL']),
      target_id: z.string().nullable(),
    }),
  ),
  /** Seconds the Agent may keep serving this bundle if Oolix is unreachable. */
  stale_grace_seconds: z.number().int().positive(),
});

export type AgentConfigBundle = z.infer<typeof AgentConfigBundleSchema>;

File: partner/agent/internal/manifest/manifest.go
// Package manifest verifies Oolix activation manifests inside the Partner
// Agent -- spec v5 §11.2, §75.
//
// This is the Go mirror of oolix/packages/manifest-schema. The two MUST agree: the
// control plane signs with one and the Agent verifies with the other, and a
// disagreement means either no ad serves at all or -- far worse -- a manifest
// the control plane considers invalid is honoured here.
//
// §11.2 states the consequence of failure plainly:
//
//	No valid signature + no valid approval + no current policy
//	= no ad and no external upload.
//
// So every function in this file fails CLOSED. There is no partial-trust path
// and no "warn and continue".
package manifest

import (
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/go-jose/go-jose/v4"
)

const (
	// JWSType is the §75 protected header typ. Checking it stops any other
	// ES256 token Oolix ever signed -- an agent access token, for instance --
	// from being accepted as a manifest.
	JWSType = "OOLIX-MANIFEST+JWS"
)

// AllowedAlgorithms pins ES256 and nothing else (§75).
//
// Passing an explicit allow-list to go-jose is what prevents algorithm
// confusion: without it a token could nominate "none" or an HMAC algorithm and
// be verified against a public key as if it were a shared secret.
var AllowedAlgorithms = []jose.SignatureAlgorithm{jose.ES256}

// Budget is the §75 budget block.
type Budget struct {
	// AllocationMinor is a DECIMAL STRING. §73 stores budgets as BIGINT and a
	// JSON number would silently lose precision past 2^53.
	AllocationMinor   string  `json:"allocation_minor"`
	Currency          string  `json:"currency"`
	LocalStopFraction float64 `json:"local_stop_fraction"`
}

// FrequencyCap is the §67.2 cap shape.
type FrequencyCap struct {
	MaxImpressions int    `json:"max_impressions"`
	Window         string `json:"window"`
}

// AudienceRule is one condition of an approved Audience Group (v6 §6.3).
//
// Duplicated here rather than imported from internal/audience so the manifest
// package keeps depending on nothing: it is the trust boundary, and the fewer
// packages that can influence what "a verified manifest" means, the better.
type AudienceRule struct {
	Attribute string      `json:"attribute"`
	Operator  string      `json:"operator"`
	Value     interface{} `json:"value"`
	Required  bool        `json:"required"`
	Weight    int         `json:"weight"`
}

// AudienceBlock is v6 Appendix B's manifest addition.
//
// RuleHash is the load-bearing field. §10 binds the Partner's approval to it,
// so the Agent recomputes it from Rules before compiling anything: if the two
// disagree the audience changed after the Partner agreed to it, and nothing may
// be served.
type AudienceBlock struct {
	AudienceGroupID   string         `json:"audience_group_id"`
	AudienceVersion   int            `json:"audience_version"`
	RuleHash          string         `json:"rule_hash"`
	Rules             []AudienceRule `json:"rules"`
	ReachEstimateID   string         `json:"reach_estimate_id,omitempty"`
	CapabilityVersion int            `json:"capability_version,omitempty"`
	MappingVersion    int            `json:"mapping_version,omitempty"`
}

// Payload is the §75 required manifest payload.
type Payload struct {
	ManifestVersion int    `json:"manifest_version"`
	ActivationID    string `json:"activation_id"`
	PartnerOrgID    string `json:"partner_org_id"`
	// SegmentID and SegmentKey are v6 §19's legacy path: a Partner with an
	// existing prebuilt segment can still be activated. Empty when the campaign
	// targets an Audience Group instead, so neither may be assumed present.
	SegmentID  string `json:"segment_id"`
	SegmentKey string `json:"segment_key"`
	// Audience is the v6 primary targeting path. Nil on a legacy manifest;
	// §80's N-1 compatibility means an Agent in the field must accept both.
	Audience           *AudienceBlock `json:"audience,omitempty"`
	Channel            string         `json:"channel"`
	PlacementIDs       []string       `json:"placement_ids"`
	PlacementKeys      []string       `json:"placement_keys"`
	CreativeVersionIDs []string       `json:"creative_version_ids"`
	Budget             Budget         `json:"budget"`
	FrequencyCap       FrequencyCap   `json:"frequency_cap"`
	PurposeID          string         `json:"purpose_id"`
	PolicyVersion      string         `json:"policy_version"`
	AllowedCategories  []string       `json:"allowed_categories"`
	BlockedCategories  []string       `json:"blocked_categories"`
	CampaignCategory   string         `json:"campaign_category"`
	IssuedAt           time.Time      `json:"issued_at"`
	ConfigExpiresAt    time.Time      `json:"config_expires_at"`
	StartAt            time.Time      `json:"start_at"`
	EndAt              time.Time      `json:"end_at"`
	ApprovalReference  string         `json:"approval_reference"`
	AudienceExpansion  bool           `json:"audience_expansion_allowed"`
}

// VerifyOptions carries the pinned values a verification is checked against.
type VerifyOptions struct {
	Issuer   string
	Audience string
	// PartnerOrgID is this Agent's own partner id. A manifest bound to a
	// different Partner is rejected outright (§75 partner binding).
	PartnerOrgID string
	// Now is injectable so expiry behaviour is testable without sleeping.
	Now time.Time
}

var (
	ErrBadSignature  = errors.New("manifest signature verification failed")
	ErrWrongType     = errors.New("manifest has an unexpected token type")
	ErrWrongIssuer   = errors.New("manifest issuer mismatch")
	ErrWrongAudience = errors.New("manifest audience mismatch")
	ErrWrongPartner  = errors.New("manifest is bound to a different Partner")
	ErrNotYetValid   = errors.New("manifest is not yet valid")
	ErrConfigExpired = errors.New("manifest config has expired")
	ErrCampaignEnded = errors.New("campaign has ended")
	ErrMalformed     = errors.New("manifest payload is malformed")
)

// Verify checks a compact JWS manifest against a JWKS.
//
// The checks run in this order, and every one of them must pass:
//
//  1. alg is ES256 and typ is the Oolix manifest type   (algorithm confusion)
//  2. signature verifies against a key in the JWKS       (authenticity)
//  3. issuer and audience are pinned                     (cross-env replay)
//  4. partner binding matches this Agent                 (cross-tenant replay)
//  5. issued_at <= now < config_expires_at               (staleness, §75)
//  6. now < end_at                                       (§24 expired campaign)
func Verify(compact string, jwks *jose.JSONWebKeySet, opts VerifyOptions) (*Payload, error) {
	now := opts.Now
	if now.IsZero() {
		now = time.Now().UTC()
	}

	sig, err := jose.ParseSigned(compact, AllowedAlgorithms)
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrBadSignature, err)
	}
	if len(sig.Signatures) != 1 {
		// A multi-signature JWS is not something Oolix produces; accepting one
		// would mean deciding which signature "counts".
		return nil, fmt.Errorf("%w: expected exactly one signature", ErrBadSignature)
	}

	header := sig.Signatures[0].Header
	protected := sig.Signatures[0].Protected

	if typ, _ := protected.ExtraHeaders[jose.HeaderType].(string); typ != JWSType {
		return nil, fmt.Errorf("%w: %q", ErrWrongType, typ)
	}

	// Select by kid so a rotation (§75 dual-key overlap) resolves to the right
	// key rather than trying every key in the set.
	var payloadBytes []byte
	if header.KeyID != "" {
		keys := jwks.Key(header.KeyID)
		if len(keys) == 0 {
			return nil, fmt.Errorf("%w: unknown kid %q", ErrBadSignature, header.KeyID)
		}
		for _, k := range keys {
			if payloadBytes, err = sig.Verify(k); err == nil {
				break
			}
		}
	} else {
		for _, k := range jwks.Keys {
			if payloadBytes, err = sig.Verify(k); err == nil {
				break
			}
		}
	}
	if payloadBytes == nil {
		return nil, ErrBadSignature
	}

	if iss, _ := protected.ExtraHeaders["iss"].(string); iss != opts.Issuer {
		return nil, fmt.Errorf("%w: %q", ErrWrongIssuer, iss)
	}
	if aud, _ := protected.ExtraHeaders["aud"].(string); aud != opts.Audience {
		return nil, fmt.Errorf("%w: %q", ErrWrongAudience, aud)
	}

	var payload Payload
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrMalformed, err)
	}

	if opts.PartnerOrgID != "" && payload.PartnerOrgID != opts.PartnerOrgID {
		return nil, ErrWrongPartner
	}

	// v6 §19: targeting is EITHER a legacy segment or an Audience Group. A
	// manifest carrying neither has no way to say who it is for, and the
	// decision engine's membership check would have nothing to check -- which
	// on a fail-open reading would serve the campaign to every user. Rejecting
	// it here means that reading never becomes reachable.
	if payload.SegmentKey == "" && payload.Audience == nil {
		return nil, fmt.Errorf("%w: manifest has neither a segment nor an audience to target", ErrMalformed)
	}
	if payload.Audience != nil && (payload.Audience.RuleHash == "" || len(payload.Audience.Rules) == 0) {
		// §10: an audience block without rules or without the hash they were
		// approved under cannot be checked against the Partner's approval.
		return nil, fmt.Errorf("%w: audience block is missing its rules or rule hash", ErrMalformed)
	}

	if now.Before(payload.IssuedAt) {
		return nil, ErrNotYetValid
	}
	if !now.Before(payload.ConfigExpiresAt) {
		// §75: the offline cache is bounded. A signature that is still
		// cryptographically valid is NOT sufficient once config expires.
		return nil, ErrConfigExpired
	}
	if !now.Before(payload.EndAt) {
		// §24: "Agent automatically stops serving even if Oolix is temporarily
		// unreachable" once the campaign end passes.
		return nil, ErrCampaignEnded
	}

	return &payload, nil
}

// ServableAt reports whether an already-verified manifest may still serve.
//
// Verification happens once when a manifest is pulled; this is re-checked on
// every decision, because a cached manifest expires while it sits in memory.
func (p *Payload) ServableAt(now time.Time) error {
	switch {
	case now.Before(p.StartAt):
		return ErrNotYetValid
	case !now.Before(p.EndAt):
		return ErrCampaignEnded
	case !now.Before(p.ConfigExpiresAt):
		return ErrConfigExpired
	default:
		return nil
	}
}

// ServesPlacement reports whether this manifest authorises a placement key.
//
// §76.2: "every placement must be Partner-published and explicitly approved."
// The Agent matches on the keys inside the signed manifest, so a placement the
// Partner did not approve cannot be served even if it exists locally.
func (p *Payload) ServesPlacement(placementKey string) bool {
	for _, k := range p.PlacementKeys {
		if k == placementKey {
			return true
		}
	}
	return false
}

// AllowsCategory applies the Partner's category policy carried in the manifest.
//
// Evaluated locally so the rule holds even when Oolix is unreachable (§75).
// Blocked wins over allowed: a category on both lists is refused.
func (p *Payload) AllowsCategory(category string) bool {
	for _, b := range p.BlockedCategories {
		if equalFold(b, category) {
			return false
		}
	}
	if len(p.AllowedCategories) == 0 {
		return true
	}
	for _, a := range p.AllowedCategories {
		if equalFold(a, category) {
			return true
		}
	}
	return false
}

func equalFold(a, b string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := 0; i < len(a); i++ {
		ca, cb := a[i], b[i]
		if 'A' <= ca && ca <= 'Z' {
			ca += 'a' - 'A'
		}
		if 'A' <= cb && cb <= 'Z' {
			cb += 'a' - 'A'
		}
		if ca != cb {
			return false
		}
	}
	return true
}
